presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkNativeTaskRdd.java

```diff
@@ -0,0 +1,185 @@
+        ImmutableList.Builder<RDD<Tuple2<MutablePartitionId, PrestoSparkMutableRow>>> shuffleInputRdds = ImmutableList.builder();
+        for (Map.Entry<String, RDD<Tuple2<MutablePartitionId, PrestoSparkMutableRow>>> entry : shuffleInputRddMap.entrySet()) {
+            shuffleInputFragmentIds.add(entry.getKey());
+            shuffleInputRdds.add(entry.getValue());
+        }
+        return new PrestoSparkNativeTaskRdd<>(context, taskSourceRdd, shuffleInputFragmentIds.build(), shuffleInputRdds.build(), taskProcessor);
+    }
+
+    @Override
+    public Iterator<Tuple2<MutablePartitionId, T>> compute(Partition split, TaskContext context)
+    {
+        PrestoSparkTaskSourceRdd taskSourceRdd = getTaskSourceRdd();
+        List<Partition> partitions = seqAsJavaList(((ZippedPartitionsPartition) split).partitions());
+        int expectedPartitionsSize = (taskSourceRdd != null ? 1 : 0) + getShuffleInputRdds().size();
+        if (partitions.size() != expectedPartitionsSize) {
+            throw new IllegalStateException(format("Unexpected partitions size. Expected: %s. Actual: %s.", expectedPartitionsSize, partitions.size()));
+        }
+
+        Iterator<SerializedPrestoSparkTaskSource> taskSourceIterator;
+        if (taskSourceRdd != null) {
+            taskSourceIterator = taskSourceRdd.iterator(partitions.get(partitions.size() - 1), context);
+        }
+        else {
+            taskSourceIterator = emptyScalaIterator();
+        }
+
+        return getTaskProcessor().process(
+                taskSourceIterator,
+                getShuffleReadDescriptors(partitions),
+                getShuffleWriteDescriptor(split));
+    }
+
+    private PrestoSparkNativeTaskRdd(
+            SparkContext context,
+            Optional<PrestoSparkTaskSourceRdd> taskSourceRdd,
+            List<String> shuffleInputFragmentIds,
+            List<RDD<Tuple2<MutablePartitionId, PrestoSparkMutableRow>>> shuffleInputRdds,
+            PrestoSparkTaskProcessor<T> taskProcessor)
+    {
+        super(context, taskSourceRdd, shuffleInputFragmentIds, shuffleInputRdds, taskProcessor);
+    }
+
+    private Map<String, PrestoSparkShuffleReadDescriptor> getShuffleReadDescriptors(List<Partition> partitions)
+    {
+        //The classloader_interface package tries to have minimal external dependencies (except the spark-core), so we use HashMap instead of Guava's ImmutableMap
+        ImmutableMap.Builder<String, PrestoSparkShuffleReadDescriptor> shuffleReadDescriptors = ImmutableMap.builder();
+
+        // Get shuffle information from ShuffledRdds for shuffle read
+        int numPartitions = partitions.size();
+        List<RDD<Tuple2<MutablePartitionId, PrestoSparkMutableRow>>> shuffleInputRdds = getShuffleInputRdds();
+        List<String> shuffleInputFragmentIds = getShuffleInputFragmentIds();
+        checkArgument(
+                numPartitions >= shuffleInputRdds.size() && numPartitions >= shuffleInputFragmentIds.size(),
+                format(
```

Comment:
format pays extra cost to format, even when there is no error.

use the checkArgument( condition, "error message %s some things %s", param1, param2) 

Note, unlike format, it only supports %s format specifiier.